### PR TITLE
Do not use PatternSet in :snapshots

### PR DESCRIPTION
**Pull request by wolfs**
Make the strategy pluggable.

### CODE REVIEW

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/PatternFilterStrategy.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+package org.gradle.internal.snapshot;
+
+import javax.annotation.Nullable;
+import java.nio.file.Path;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.function.BiPredicate;
+
+public interface PatternFilterStrategy<PATTERN> {
```
Comment (wolfs) [inline subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/PatternFilterStrategy.java]: Better name wanted

File: subprojects/snapshots/snapshots.gradle.kts [hunk 2]
```diff
@@ -27,7 +27,6 @@ dependencies {
     implementation(project(":native"))
     implementation(project(":messaging"))
     implementation(project(":coreApi"))
-    implementation(project(":files"))
```
Comment (lptr) [inline subprojects/snapshots/snapshots.gradle.kts]: \o/

File: subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/PatternSetSnapshottingFilter.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.function.BiPredicate;
+
+public class PatternSetSnapshottingFilter implements SnapshottingFilter {
+    private final PatternSet patternSet;
+    private final FileSystem fileSystem;
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/PatternSetSnapshottingFilter.java]: We only need `Stat` here. It looks like that's what we need to pass around instead of `FileSystem`.

**Comment by wolfs** [comment]
@lptr PTAL!

File: subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/PatternSetSnapshottingFilter.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+        this.patternSet = patternSet;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return patternSet.isEmpty();
+    }
+
+    @Override
+    public BiPredicate<FileSystemLocationSnapshot, Iterable<String>> getAsSnapshotPredicate() {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/PatternSetSnapshottingFilter.java]: We could pass in `Stat` here instead of storing it as a field. Not sure if it's better, but maybe?

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DefaultFileSystemSnapshotter.java [hunk 5]
```diff
@@ -182,15 +181,14 @@ public FileSystemSnapshotBuilder newFileSystemSnapshotBuilder() {
         return new FileSystemSnapshotBuilder(stringInterner, hasher);
     }
 
-    private FileSystemSnapshot filterSnapshot(FileSystemLocationSnapshot snapshot, PatternSet patterns) {
+    private FileSystemSnapshot filterSnapshot(FileSystemLocationSnapshot snapshot, SnapshottingFilter snapshottingFilter) {
         if (snapshot.getType() == FileType.Missing) {
             return FileSystemSnapshot.EMPTY;
         }
-        if (patterns.isEmpty()) {
+        if (snapshottingFilter.isEmpty()) {
```
Comment (lptr) [inline subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DefaultFileSystemSnapshotter.java]: Nit: calling these parameters `filter` should be fine, too.

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java [hunk 6]
```diff
@@ -105,15 +91,15 @@ public DefaultExcludes(String[] defaultExcludes) {
                     if (firstStar == -1) {
                         excludeFiles.add(defaultExclude);
                     } else {
-                        Predicate<String> start = firstStar == 0 ? Predicates.<String>alwaysTrue() : new StartMatcher(defaultExclude.substring(0, firstStar));
-                        Predicate<String> end = firstStar == length - 1 ? Predicates.<String>alwaysTrue() : new EndMatcher(defaultExclude.substring(firstStar + 1, length));
-                        excludeFileSpecs.add(Predicates.and(start, end));
+                        Predicate<String> start = firstStar == 0 ? it -> true : new StartMatcher(defaultExclude.substring(0, firstStar));
+                        Predicate<String> end = firstStar == length - 1 ? it -> true : new EndMatcher(defaultExclude.substring(firstStar + 1, length));
```
Comment (lptr) [inline subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java]: Should we add parenthesis around these guys? Looks weird to me naked like this.

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/SnapshottingFilter.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        @Override
+        public DirectoryWalkerPredicate getAsDirectoryWalkerPredicate() {
+            return (path, name, isDirectory, attrs, relativePath) -> true;
+        }
+    };
+
+    boolean isEmpty();
+    BiPredicate<FileSystemLocationSnapshot, Iterable<String>> getAsSnapshotPredicate();
```
Comment (lptr) [inline subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/SnapshottingFilter.java]: Can we please have a custom type for this guy as well, like with `DirectoryWalkerPredicate`?

File: subprojects/version-control/version-control.gradle.kts [hunk 8]
```diff
@@ -39,6 +39,7 @@ dependencies {
     implementation(library("jsch"))
 
     testImplementation(project(":native"))
+    testImplementation(project(":snapshots"))
```
Comment (lptr) [inline subprojects/version-control/version-control.gradle.kts]: Why is this needed?

Comment (wolfs) [reply subprojects/version-control/version-control.gradle.kts]: We are using `TestFiles` in those projects, and that type now depends on `FileSystemSnapshotter`. Moreover, over test fixtures don't do transitive dependencies :(

So adding the dependency seems to be the current workaround (see [this Slack discussion](https://gradle.slack.com/archives/C03A0MT6S/p1560348413072200?thread_ts=1560339024.071900&cid=C03A0MT6S)).

File: subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/PatternSetSnapshottingFilter.java (hunk 4 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/PatternSetSnapshottingFilter.java]: I like that we don't need `Stat` for an arbitrary `SnapshottingFilter`, but only when creating one for `PatternSpec`s. The `DefaultFileSystemSnapshotter` also needs something different to `Stat`, given that we don't have the class available.

File: subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/PatternSetSnapshottingFilter.java (hunk 3 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/internal/fingerprint/impl/PatternSetSnapshottingFilter.java]: I am now moving around `Stat`. Note that `AbstractFileTreeElement` needed `Chmod`. I don't extend `AbstractFileTreeElement` any more and directly implement `FileTreeElement`, so there is no need for `Chmod`.

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DefaultFileSystemSnapshotter.java (hunk 5 above)
Comment (wolfs) [reply subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DefaultFileSystemSnapshotter.java]: Done.

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java (hunk 6 above)
Comment (wolfs) [reply subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java]: I changed the formatting.

File: subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/SnapshottingFilter.java (hunk 7 above)
Comment (wolfs) [reply subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/SnapshottingFilter.java]: Introduced `FileSystemSnapshotPredicate`.

**Comment by wolfs** [comment]
@lptr PTAL!
